src/ATokenNaive.sol

```diff
@@ -0,0 +1,357 @@
+  }
+
+  /// TODO how to handle onBehalfOf?
+  /// TODO should this revert if the vote has been cast?
+  /// @notice Allow a depositor to express their voting preference for a given
+  /// proposal. Their preference is recorded internally but not moved to the
+  /// Governor until `castVote` is called. We deliberately do NOT revert if the
+  /// internalVotingPeriodEnd has passed.
+  /// @param proposalId The proposalId in the associated Governor
+  /// @param support The depositor's vote preferences in accordance with the `VoteType` enum.
+  function expressVote(uint256 proposalId, uint8 support) external {
+    uint256 weight = getPastDeposits(msg.sender, governor.proposalSnapshot(proposalId));
+    if (weight == 0) revert("no weight");
+
+    if (_proposalVotersHasVoted[proposalId][msg.sender]) revert("already voted");
+    _proposalVotersHasVoted[proposalId][msg.sender] = true;
+
+    if (support == uint8(VoteType.Against)) {
+      proposalVotes[proposalId].againstVotes += SafeCast.toUint128(weight);
+    } else if (support == uint8(VoteType.For)) {
+      proposalVotes[proposalId].forVotes += SafeCast.toUint128(weight);
+    } else if (support == uint8(VoteType.Abstain)) {
+      proposalVotes[proposalId].abstainVotes += SafeCast.toUint128(weight);
+    } else {
+      revert("invalid support value, must be included in VoteType enum");
+    }
+  }
+
+  /// @notice Causes this contract to cast a vote to the Governor for all the
+  /// tokens it currently holds.  Uses the sum of all depositor voting
+  /// expressions to decide how to split its voting weight. Can be called by
+  /// anyone, but _must_ be called within `CAST_VOTE_WINDOW` blocks before the
+  /// proposal deadline.
+  /// @param proposalId The ID of the proposal which the Pool will now vote on.
+  function castVote(uint256 proposalId) external {
+    if (internalVotingPeriodEnd(proposalId) > block.number) revert("cannot castVote yet");
+    uint8 unusedSupportParam = uint8(VoteType.Abstain);
+    ProposalVote memory _proposalVote = proposalVotes[proposalId];
+
+    uint256 _proposalSnapshotBlockNumber = governor.proposalSnapshot(proposalId);
+
+    // Use the snapshot of total deposits to determine total voting weight. We cannot
+    // use the proposalVote numbers alone, since some people with deposits at the
+    // snapshot might not have expressed votes.
+    uint256 _totalDepositWeightAtSnapshot = getPastTotalDeposits(_proposalSnapshotBlockNumber);
+
+    // We need 256 bits because of the multiplication we're about to do.
+    uint256 _votingWeightAtSnapshot = IVotingToken(address(_underlyingAsset)).getPastVotes(
+      address(this), _proposalSnapshotBlockNumber
+    );
+
+    //      forVotesRaw          forVotesScaled
+    // --------------------- = ---------------------
+    //     totalDeposits        deposits (@snapshot)
+    //
+    // forVotesScaled = forVotesRaw * deposits@snapshot / totalDeposits
```

Comment:
Is this correct: we are forced to vote with the full voting weight, so we need to take the counts and scale them up to the total number of votes since some users may not have voted?

Reply:
> Is this correct: we are forced to vote with the full voting weight,

No, we aren't forced to. We could vote with less. I think we felt it was just a waste of the votes to not use the entire weight if we could.

Reply:
Then what is the scaling / why isn't it just "submit the exact amount of expressed votes"? The comments say:

> We cannot use the proposalVote numbers alone, since some people with deposits at the snapshot might not have expressed votes.

But if they didn't express votes, wouldn't be in the `proposalVotes` mapping?

Reply:
We have to scale because people might have borrowed GOV, so the actual voting weight of the aToken is probably less than the number of deposits. So expressed votes have to be scaled down

Reply:
In other words, the scaling is: "Vote with the tokens left in the pool proportional to your AToken holdings"

Reply:
Ok, I get this now. Had to think about it bit more but now I see. Thanks!